objects/client.c

```diff
@@ -1424,6 +1455,41 @@ client_manage(xcb_window_t w, xcb_get_geometry_reply_t *wgeom, xcb_get_window_at
                           globalconf.default_cmap
                       });
 
+    /* is a container window needed? */
+    c->container_window = XCB_NONE;
+    if (wgeom->depth != globalconf.default_depth) {
+        /* The window has a different depth than the window that we would put it
+         * in. If it also has a ParentRelative background, ReparentWindow would
+         * fail. We cannot check for that, so assume the worst.
+         */
+        xcb_colormap_t colormap;
+        xcb_visualid_t visual = pick_visual_for_depth(wgeom->depth);
+
+        if (wgeom->depth == globalconf.screen->root_depth)
+            colormap = globalconf.screen->default_colormap;
+        else
+        {
+            colormap = xcb_generate_id(globalconf.connection);
+            xcb_create_colormap(globalconf.connection, XCB_COLORMAP_ALLOC_NONE,
+                    colormap, globalconf.screen->root, visual);
+        }
+
+        c->container_window = xcb_generate_id(globalconf.connection);
+        xcb_create_window(globalconf.connection, wgeom->depth, c->container_window, c->frame_window,
+                0, 0, wgeom->width, wgeom->height, 0, XCB_COPY_FROM_PARENT,
+                visual, XCB_CW_BACK_PIXMAP | XCB_CW_BORDER_PIXEL | XCB_CW_EVENT_MASK | XCB_CW_COLORMAP,
+                (const uint32_t[]) {
+                    XCB_BACK_PIXMAP_NONE,
+                    globalconf.screen->black_pixel,
+                    XCB_EVENT_MASK_SUBSTRUCTURE_REDIRECT,
+                    colormap
+                });
+        xcb_map_window(globalconf.connection, c->container_window);
+
+        if (colormap != globalconf.screen->default_colormap)
+            xcb_free_colormap(globalconf.connection, colormap);
```

Comment:
I see you free the colormap before the window is destroyed. Is it really allowed?

https://tronche.com/gui/x/xlib/color/XFreeColormap.html says the following:
> If the specified colormap is defined as the colormap for a window (by XCreateWindow(), [...]), XFreeColormap() changes the colormap associated with the window to None [...]. X does not define the colors displayed for a window with a colormap of None.

Reply:
Well, it is definitely allowed, since you linked to documentation explaining what happens. ;-)

However, a window with colormap `None` and a depth different from its parent is indeed... "weird".

To be honest, I was expecting this to be similar to creating a pixmap, creating a render picture, and then destroying the pixmap. The picture keeps the pixmap alive for as long as needed.